Properly include declaration of dispatch in file that registers it.

ezyang opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #30315 Split libtorch.so back into libtorch_{cpu,cuda,hip}
* #30314 Add missing trigramma_stub definition.
* #30313 Workaround hcc bug regarding extern "C" definitions
* #30312 Delete redundant THC_API on THCStorage_new
* **#30311 Properly include declaration of dispatch in file that registers it.**
* #30310 Add missing _API definitions.
* #30308 DEFINE_DISPATCH in the correct namespace.
* #30307 Use normal dispatch to get to CUDA threshold kernels, instead of DispatchStub.

multinomial_stub must be in scope to register against it.  Somehow,
this works today, but when I split torch_cpu and torch_cuda it
doesn't.

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

Differential Revision: [D18762358](https://our.internmc.facebook.com/intern/diff/D18762358)

ezyang committed:
Properly include declaration of dispatch in file that registers it.

multinomial_stub must be in scope to register against it.  Somehow,
this works today, but when I split torch_cpu and torch_cuda it
doesn't.

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

[ghstack-poisoned]

ezyang committed:
Update on "Properly include declaration of dispatch in file that registers it."


multinomial_stub must be in scope to register against it.  Somehow,
this works today, but when I split torch_cpu and torch_cuda it
doesn't.

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

[ghstack-poisoned]

ezyang committed:
Update on "Properly include declaration of dispatch in file that registers it."


multinomial_stub must be in scope to register against it.  Somehow,
this works today, but when I split torch_cpu and torch_cuda it
doesn't.

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

Differential Revision: [D18762358](https://our.internmc.facebook.com/intern/diff/D18762358)

[ghstack-poisoned]

ezyang committed:
Update on "Properly include declaration of dispatch in file that registers it."


multinomial_stub must be in scope to register against it.  Somehow,
this works today, but when I split torch_cpu and torch_cuda it
doesn't.

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

Differential Revision: [D18762358](https://our.internmc.facebook.com/intern/diff/D18762358)

[ghstack-poisoned]

albanD reviewed (commented):
Why is that needed? It wasn't working before?

ezyang commented:
> Why is that needed? It wasn't working before?

I'm not sure! But it seems obviously correct to include the header, so I didn't investigate further.

ezyang committed:
Update on "Properly include declaration of dispatch in file that registers it."


multinomial_stub must be in scope to register against it.  Somehow,
this works today, but when I split torch_cpu and torch_cuda it
doesn't.

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

Differential Revision: [D18762358](https://our.internmc.facebook.com/intern/diff/D18762358)

[ghstack-poisoned]
